File: mdm-opsflow-phase2-starter/backend/app/api/routes/document_intake.py
from __future__ import annotations

from fastapi import APIRouter, Depends, File, UploadFile, status
from sqlalchemy.orm import Session

from app.db import get_db
from app.dependencies import RequestContext, require_permissions, resolve_tenant_scope
from app.models import AuditLog
from app.schemas import DocumentIntakeConfigResponse, DocumentIntakeProjectResponse, DocumentIntakeRouteResponse, DocumentIntakeVendorResponse
from app.services.document_intake_router import (
    AUTO_POST_FINANCIAL_OR_TICKET_MIN_CONFIDENCE,
    AUTO_ROUTE_MIN_CONFIDENCE,
    DOCUMENT_RULES,
    UNKNOWN_ROUTE,
    route_ocr_document,
)
from app.services.intake_processing import extract_document_text


router = APIRouter(prefix="/api/document-intake", tags=["Document Intake"])


def _routes_config() -> dict[str, str]:
    routes = {document_type: route for document_type, (_, route, _) in DOCUMENT_RULES.items()}
    routes["unknown"] = UNKNOWN_ROUTE
    return routes


@router.get(
    "/config",
    response_model=DocumentIntakeConfigResponse,
    operation_id="document_intake_config_get",
    summary="Get document intake routing config",
    description="Returns confidence gates and supported document routing destinations for OCR document intake.",
)
def get_document_intake_config(
    context: RequestContext = Depends(require_permissions("intake_read")),
) -> DocumentIntakeConfigResponse:
    _ = resolve_tenant_scope(context)
    return DocumentIntakeConfigResponse(
        auto_route_min_confidence=AUTO_ROUTE_MIN_CONFIDENCE,
        auto_post_financial_or_ticket_min_confidence=AUTO_POST_FINANCIAL_OR_TICKET_MIN_CONFIDENCE,
        never_silent_overwrite=True,
        preserve_source_value=True,
        preserve_units=True,
        flag_cross_document_conflicts=True,
        require_tenant_scope=True,
        create_audit_event=True,
        routes=_routes_config(),
    )


@router.post(
    "",
    response_model=DocumentIntakeRouteResponse,
    status_code=status.HTTP_200_OK,
    operation_id="document_intake_route_upload",
    summary="Classify and route an uploaded document",
    description=(
        "Extracts OCR text from an uploaded PDF, image, or text file, classifies the document, "
        "extracts supported facts only, and returns strict routing JSON for review/import screens."
    ),
)
async def route_uploaded_document(
    file: UploadFile = File(...),
    context: RequestContext = Depends(require_permissions("intake_write")),
    db: Session = Depends(get_db),
) -> DocumentIntakeRouteResponse:
    tenant_id = resolve_tenant_scope(context)
    payload = await file.read()
    raw_text = extract_document_text(payload, file.content_type or "application/octet-stream")
    result = route_ocr_document(raw_text)

    db.add(
        AuditLog(
            tenant_id=tenant_id,
            actor_user_id=context.user.id,
            action="document_intake_route_preview",
            resource_type="document_intake",
            resource_id=file.filename or "upload",
            details=f"document_type={result.document_type}; route={result.recommended_route}; confidence={result.classification_confidence}",
            created_by=context.user.id,
        )
    )
    db.commit()

    return DocumentIntakeRouteResponse(
        document_type=result.document_type,
        classification_confidence=result.classification_confidence,
        recommended_route=result.recommended_route,
        project=DocumentIntakeProjectResponse(**result.project),
        vendor=DocumentIntakeVendorResponse(**result.vendor),
        extracted_fields=result.extracted_fields,
        uncertain_fields=result.uncertain_fields,
        conflicts=result.conflicts,
        requires_human_review=result.requires_human_review,
        reason_for_review=result.reason_for_review,
    )

File: mdm-opsflow-phase2-starter/frontend/lib/documentIntake.ts
import { clearSession, getAccessToken, getTenantId, refreshSession } from "@/lib/auth";
import { getApiBaseUrl } from "@/lib/i18n";

export type DocumentIntakeConfig = {
  auto_route_min_confidence: number;
  auto_post_financial_or_ticket_min_confidence: number;
  never_silent_overwrite: boolean;
  preserve_source_value: boolean;
  preserve_units: boolean;
  flag_cross_document_conflicts: boolean;
  require_tenant_scope: boolean;
  create_audit_event: boolean;
  routes: Record<string, string>;
};

export type DocumentIntakeRouteResult = {
  document_type: string;
  classification_confidence: number;
  recommended_route: string;
  project: { name: string | null; number: string | null; match_confidence: number };
  vendor: { name: string | null; document_number: string | null };
  extracted_fields: Record<string, unknown>;
  uncertain_fields: string[];
  conflicts: Array<Record<string, unknown>>;
  requires_human_review: boolean;
  reason_for_review: string | null;
};

export class DocumentIntakeApiError extends Error {
  status: number;
  detail: string;

  constructor(status: number, detail: string) {
    super(detail);
    this.status = status;
    this.detail = detail;
  }
}

function buildAuthHeaders(): Record<string, string> {
  const token = getAccessToken();
  const tenantId = getTenantId();
  if (!token) {
    throw new Error("Missing auth token");
  }
  if (!tenantId) {
    throw new Error("Missing tenant context");
  }

  return {
    Authorization: `Bearer ${token}`,
    "X-Tenant-ID": tenantId,
  };
}

async function parseError(response: Response, fallback: string): Promise<DocumentIntakeApiError> {
  let detail = fallback;
  try {
    const payload = (await response.json()) as { detail?: string };
    if (payload?.detail) {
      detail = payload.detail;
    }
  } catch {
    // Keep fallback for non-JSON errors.
  }
  return new DocumentIntakeApiError(response.status, detail);
}

async function fetchWithSessionRetry(input: string, init: RequestInit, fallback: string): Promise<Response> {
  const apiBaseUrl = getApiBaseUrl();
  let response = await fetch(input, { ...init, headers: { ...buildAuthHeaders(), ...(init.headers as Record<string, string> | undefined) } });

  if (response.status === 401 && await refreshSession(apiBaseUrl)) {
    response = await fetch(input, { ...init, headers: { ...buildAuthHeaders(), ...(init.headers as Record<string, string> | undefined) } });
  }

  if (response.status === 401) {
    clearSession();
  }

  if (!response.ok) {
    throw await parseError(response, fallback);
  }

  return response;
}

export async function getDocumentIntakeConfig(): Promise<DocumentIntakeConfig> {
  const response = await fetchWithSessionRetry(
    `${getApiBaseUrl()}/api/document-intake/config`,
    { method: "GET" },
    "Unable to load document intake configuration."
  );
  return (await response.json()) as DocumentIntakeConfig;
}

export async function routeDocumentForIntake(file: File): Promise<DocumentIntakeRouteResult> {
  const formData = new FormData();
  formData.append("file", file);

  const response = await fetchWithSessionRetry(
    `${getApiBaseUrl()}/api/document-intake`,
    { method: "POST", body: formData },
    "Unable to route document for intake."
  );
  return (await response.json()) as DocumentIntakeRouteResult;
}